File: src/main/java/com/baulsupp/oksocial/services/sheetsu/SheetsuAuthInterceptor.java
package com.baulsupp.oksocial.services.sheetsu;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.authenticator.BasicCredentials;
import com.baulsupp.oksocial.authenticator.ValidatedCredentials;
import com.baulsupp.oksocial.credentials.ServiceDefinition;
import com.baulsupp.oksocial.output.OutputHandler;
import com.baulsupp.oksocial.secrets.Secrets;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;
import okhttp3.Credentials;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class SheetsuAuthInterceptor implements AuthInterceptor<BasicCredentials> {
  @Override public SheetsuServiceDefinition serviceDefinition() {
    return new SheetsuServiceDefinition();
  }

  @Override
  public Response intercept(Interceptor.Chain chain, BasicCredentials credentials)
      throws IOException {
    Request request = chain.request();

    request =
        request.newBuilder()
            .addHeader("Authorization", Credentials.basic(credentials.user, credentials.password))
            .build();

    return chain.proceed(request);
  }

  @Override public Future<Optional<ValidatedCredentials>> validate(OkHttpClient client,
      Request.Builder requestBuilder, BasicCredentials credentials) throws IOException {
    return CompletableFuture.completedFuture(Optional.empty());
  }

  @Override public BasicCredentials authorize(OkHttpClient client, OutputHandler outputHandler,
      List<String> authArguments) {
    String user =
        Secrets.prompt("Sheetsu API Key", "sheetsu.apiKey", "", false);
    String password =
        Secrets.prompt("Sheetsu API Password", "sheetsu.apiSecret", "", true);

    return new BasicCredentials(user, password);
  }

  @Override public Collection<String> hosts() {
    return SheetsuUtil.API_HOSTS;
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/services/sheetsu/SheetsuAuthInterceptor.kt
package com.baulsupp.oksocial.services.sheetsu

import com.baulsupp.oksocial.authenticator.AuthInterceptor
import com.baulsupp.oksocial.authenticator.BasicCredentials
import com.baulsupp.oksocial.authenticator.basic.BasicAuthServiceDefinition
import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.secrets.Secrets
import okhttp3.Credentials
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response

class SheetsuAuthInterceptor : AuthInterceptor<BasicCredentials>() {
  override fun serviceDefinition(): BasicAuthServiceDefinition =
    BasicAuthServiceDefinition("sheetsu.com", "Sheetsu API", "sheetsu", "https://docs.sheetsu.com/", "https://sheetsu.com/dashboard")

  override fun intercept(chain: Interceptor.Chain, credentials: BasicCredentials): Response {
    var request = chain.request()

    request = request.newBuilder()
      .addHeader("Authorization", Credentials.basic(credentials.user, credentials.password))
      .build()

    return chain.proceed(request)
  }

  suspend override fun authorize(client: OkHttpClient, outputHandler: OutputHandler<Response>, authArguments: List<String>): BasicCredentials {
    val user =
      Secrets.prompt("Sheetsu API Key", "sheetsu.apiKey", "", false)
    val password =
      Secrets.prompt("Sheetsu API Password", "sheetsu.apiSecret", "", true)

    return BasicCredentials(user, password)
  }

  override fun hosts(): Set<String> = setOf("sheetsu.com")
}
